python_benchmarks/benchmark_layernorm.py

```diff
@@ -0,0 +1,148 @@
+from nvfuser import FusionDefinition, DataType
+from nvfuser.pytorch_utils import torch_dtype_to_nvfuser_dtype
+from .core import runBenchmark
+import torch
+from .global_params import pytestmark, RTOL, ATOL
+
+
+def layernorm_fwd_fusion(
+    fd: FusionDefinition,
+    dtype: DataType,
+    eps: float = 1e-5,
+) -> None:
+    T0 = fd.define_tensor(
+        shape=[-1, -1], contiguity=[True, True], dtype=dtype, is_cpu=False
+    )
+
+    if dtype is not DataType.Float:
+        T0 = fd.ops.cast(T0, dtype=DataType.Float)
+
+    T2, T3 = fd.ops.var_mean(T0, axes=[1], correction=0, keepdim=False)
+
+    V6 = fd.define_vector([T0.size(0), 1], dtype=DataType.Int)
+    T7 = fd.ops.broadcast_in_dim(T2, shape=V6, broadcast_dims=[0])
+    T11 = fd.ops.broadcast_in_dim(T3, shape=V6, broadcast_dims=[0])
+
+    S12 = fd.define_scalar(eps, dtype=DataType.Double)
```

Comment:
Should this be `dtype=dtype` because you already passed in the desired dtype from outside?

Reply:
If you see at the beginning of the definition, the inputs are all casted to `Datatype.Float`. The desired `dtype` makes sure the output is cast back to the original datatype of the inputs. Irrespective of the input dtype, the `thunder` generated fusion definitions cast the inputs to `Float` and then cast back to original type, and defined this scalar as `Datatype.Double`.

Reply:
It should only promote and not downcast though right? If dtype is Double, will we still cast to Float? I believe casting to float is only necessary for reduced precision types which do not have arithmetic support in cuda.

Reply:
There are two datatypes here:

1. Argument `dtype`: This is the datatype of the incoming inputs. If the inputs are not Float, they are casted to Float for the computation and before returning, the outputs are casted back to the original dtype. This is necessary for the reduced precision types as @jacobhinkle mentioned.

2. Datatype of scalars declared within the fusion definition: The argument `dtype` is unrelated to the dtype used for any scalars in the definition. In this particular example, `thunder` declares the scalar constant variable `eps` to be of type `Double` regardless of the argument `dtype`. Similarly other scalars in `layernorm` are also of type `Double`. These scalars are not inputs to the fusion definitions but constants specific to the fusion.

Reply:
I misunderstood earlier. I generated the fusion definitions for `bfloat16`, `float16` and `float32` and used those to define a combined fusion definition. We did not have `double` datatype in cpp benchmarks. I'll check the definition from `thunder` for this case.

Reply:
Switched this to only cast for dtypes in `PROMOTE_DTYPES = [bfloat16 ,float16]`. Although, we are currently not benchmarking`double`.